pulp_rpm/tests/functional/api/test_domains.py

```diff
@@ -0,0 +1,533 @@
+    rpm_package_factory,
+    gen_object_with_cleanup,
+    rpm_package_api,
+):
+    """Test uploading of RPM content with domains."""
+    body = {
+        "name": str(uuid.uuid4()),
+        "storage_class": "pulpcore.app.models.storage.FileSystem",
+        "storage_settings": {"MEDIA_ROOT": "/var/lib/pulp/media/"},
+    }
+    domain = gen_object_with_cleanup(domains_api_client, body)
+
+    try:
+        default_content = rpm_package_factory()
+        domain_content = rpm_package_factory(pulp_domain=domain.name)
+        assert default_content.pulp_href != domain_content.pulp_href
+        assert default_content.sha256 == domain_content.sha256
+
+        domain_contents = rpm_package_api.list(pulp_domain=domain.name)
+        assert domain_contents.count == 1
+    finally:
+        cleanup_domains([domain], content_api_client=rpm_package_api)
+
+
+@pytest.mark.parallel
+def test_content_promotion(
+    cleanup_domains,
+    domains_api_client,
+    download_content_unit,
+    rpm_repository_api,
+    rpm_rpmremote_factory,
+    rpm_publication_api,
+    rpm_distribution_factory,
+    gen_object_with_cleanup,
+    monitor_task,
+):
+    """Tests Content promotion path with domains: Sync->Publish->Distribute"""
+    body = {
+        "name": str(uuid.uuid4()),
+        "storage_class": "pulpcore.app.models.storage.FileSystem",
+        "storage_settings": {"MEDIA_ROOT": "/var/lib/pulp/media/"},
+    }
+    domain = gen_object_with_cleanup(domains_api_client, body)
+
+    try:
+        # Sync task
+        remote = rpm_rpmremote_factory(pulp_domain=domain.name)
+        repo_body = {"name": str(uuid.uuid4()), "remote": remote.pulp_href}
+        repo = rpm_repository_api.create(repo_body, pulp_domain=domain.name)
+
+        task = rpm_repository_api.sync(repo.pulp_href, {}).task
+        response = monitor_task(task)
+        assert len(response.created_resources) == 1
+
+        repo = rpm_repository_api.read(repo.pulp_href)
+        assert repo.latest_version_href[-2] == "1"
+
+        # Publish task
+        pub_body = {"repository": repo.pulp_href}
+        task = rpm_publication_api.create(pub_body, pulp_domain=domain.name).task
+        response = monitor_task(task)
+        assert len(response.created_resources) == 1
+        pub_href = response.created_resources[0]
+        pub = rpm_publication_api.read(pub_href)
+
+        assert pub.repository == repo.pulp_href
+
+        # Distribute Task
+        distro = rpm_distribution_factory(publication=pub.pulp_href, pulp_domain=domain.name)
+
+        assert distro.publication == pub.pulp_href
+        # Url structure should be host/CONTENT_ORIGIN/DOMAIN_PATH/BASE_PATH
+        assert domain.name == distro.base_url.rstrip("/").split("/")[-2]
+
+        # Check that content can be downloaded from base_url
+        for pkg in ("bear-4.1-1.noarch.rpm", "pike-2.2-1.noarch.rpm"):
+            pkg_path = get_package_repo_path(pkg)
+            download_content_unit(distro.base_path, pkg_path)
```

Comment:
I still don't think this works, you need to pass in the domain name to this function else it will try to download it from the default domain. https://github.com/pulp/pulpcore/blob/main/pulpcore/tests/functional/__init__.py#L986

Reply:
Ah damn, You Are Correct Sir! Good catch.